### ISSUE TITLE
How to format results on testing set and submit for lane approximations task?

### ISSUE THREAD

**Issue by liuruijin17** [karstenBehrendt/unsupervised_llamas#10]
Thanks for your fantastic work and high-quality datasets! 

I am working on the lane approximations task on your dataset and wonder:
1. the "mean deviation" 17.8 in your paper results from the **validation set** or **test set**?
2. following the Eq. 5 in your paper, I can only get the result on the validation set. How can I format my result files, and where should I submit to evaluate performance on the **test set**.

I look forward to your response! Thanks!

**Comment by karstenBehrendt** [comment]
It looks like I didn't upload the validation scripts for that yet. Please give me a couple of days to validate them and add them to the repo. I'll let you know how to best submit the results.

The results reported within the paper are from the test set. I will make sure to update the results as part of the website as well.

**Comment by liuruijin17** [comment]
> It looks like I didn't upload the validation scripts for that yet. Please give me a couple of days to validate them and add them to the repo. I'll let you know how to best submit the results.
> 
> The results reported within the paper are from the test set. I will make sure to update the results as part of the website as well.

Thanks for your quick reply and looking forward to your updates!

**Comment by karstenBehrendt** [comment]
@liuruijin17 

https://github.com/karstenBehrendt/unsupervised_llamas/blob/master/lane_regression/evaluate.py
Here is the evaluation script. It also contains a brief description of the label file. Please let me know if you have any questions.

In the following days, I plan to upload a very simple baseline that only takes the average lane location, and maybe in a few weeks update the results from the paper based on the new evaluation to keep everything fair.  Feel free to submit your results at any time and I will speed things up considerably.

**Comment by karstenBehrendt** [comment]
I added an example and a quick way to add sample submission files 
https://github.com/karstenBehrendt/unsupervised_llamas/tree/master/lane_regression

**Comment by liuruijin17** [comment]
> @liuruijin17
> 
> https://github.com/karstenBehrendt/unsupervised_llamas/blob/master/lane_regression/evaluate.py
> Here is the evaluation script. It also contains a brief description of the label file. Please let me know if you have any questions.
> 
> In the following days, I plan to upload a very simple baseline that only takes the average lane location, and maybe in a few weeks update the results from the paper based on the new evaluation to keep everything fair. Feel free to submit your results at any time and I will speed things up considerably.

Thanks for your quick response and excellent solution! However, I still have some questions about

1. Why it is necessary to identify which lane is the l1, l0, r0, or r1? In TuSimple and CULane, there is no need to identify those subclasses.
2. What if an image has less than 4 lanes? If an image has 3 lanes, the subclasses should be "l1 l0 r0", "l0 r0 r1", or "l0 r0 r1"? What if 2 lanes in an image?

My opinion:

The subclasses should be ignored. If one image has N ground-truth lanes, and a detector gives M detected lanes. Then the evaluation script should find best matches (For a gt **G** and a det **D**,  if the number of horizontal distance < 20 pixels is more than %85 like TuSimple, that gt **G** is matched for that det **D**). The ground-truth lanes that have no matched detections are reported as False Negatives.

Last thing:

Could you please tell us the mean-deviation on the validation set in your work? We want to fine-tune our model better then submit the result files on the test set.

Thanks a lot! Your dataset and annotations are better than any other datasets!

**Comment by karstenBehrendt** [comment]
1.) It's a different kind of benchmark. Knowing where the lanes are greatly help with functionality after the lane detection. It's just a different emphasis of this benchmark.
2.) If an image has less than 2 lanes, only the annotated lanes are being evaluated. I would recommend to add "mean" lanes where your algorithm doesn't detect anything. It will break if there are annotations where you don't provide anything. Otherwise, metrics will become to complicated to compare

The main difference is that this benchmark expects lanes to be detected and mainly goes for accuracy.

However, the benchmark is open source. If you feel like adding a simpler tp / fp / fn challenge based on a pixel difference of, e.g., 20 pixels, feel free to add that to the repo. If you submit a baseline for that challenge, I'll also add it to the website under benchmarks here https://unsupervised-llamas.com/llamas/benchmarks

**Comment by liuruijin17** [comment]
> 1.) It's a different kind of benchmark. Knowing where the lanes are greatly help with functionality after the lane detection. It's just a different emphasis of this benchmark.
> 2.) If an image has less than 2 lanes, only the annotated lanes are being evaluated. I would recommend to add "mean" lanes where your algorithm doesn't detect anything. It will break if there are annotations where you don't provide anything. Otherwise, metrics will become to complicated to compare
> 
> The main difference is that this benchmark expects lanes to be detected and mainly goes for accuracy.
> 
> However, the benchmark is open source. If you feel like adding a simpler tp / fp / fn challenge based on a pixel difference of, e.g., 20 pixels, feel free to add that to the repo. If you submit a baseline for that challenge, I'll also add it to the website under benchmarks here https://unsupervised-llamas.com/llamas/benchmarks

Thanks for your reply! I now understand and follow your benchmark. The evaluation script works for the validation set.

Now I want to submit my test file but I encounter "413 Request Entity Too Large" after uploading the whole file. My original test result file in .json is ~1.01GB and I compress it to .zip which is ~484MB but still I face the "413 Request Entity Too Large" problem. So, 

1. Should I send the result file by sending you an e-mail?
2. Do I need to add all the basic information about my entry in that e-mail? (like the software environment, username, etc.)
3. What is the correct **_e-mail address_** that I should send to?

Thank you so much for taking the time to seriously respond to my question!

**Comment by karstenBehrendt** [comment]
I updated the submission form with https://github.com/karstenBehrendt/benchmarks_website/commit/d0c6d0d1368840a1cf079c970a737f6b0a4e8ddf and the submission form is still available at https://unsupervised-llamas.com/llamas/submission

Please just enter all of your information there including a link to the results file. You will need to upload your results file to dropbox, google drive, your own server or something. Unfortunately, my server regularly timed out during uploads which is why I'm now moving to that option.

You can also email me directly <llamas  (at)  kbehrendt (dot) com> but please don't send me results files. My mail server will not be able to handle those.

**Comment by karstenBehrendt** [comment]
(Only a subset of the values within the submission page are mandatory, i.e., model name, model url, inference speed)
